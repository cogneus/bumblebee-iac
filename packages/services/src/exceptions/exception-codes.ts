import { ExceptionCodes } from 'bumblebee-common'

export const TemplateExceptionCodes = {
  ...ExceptionCodes,
  TemplateNotFound: 'TemplateNotFound',
}
