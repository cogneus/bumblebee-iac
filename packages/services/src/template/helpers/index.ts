export * from './keys.helper'
export * from './validate-template.helper'
export * from './generate-id'
export * from './filter-input.class'
