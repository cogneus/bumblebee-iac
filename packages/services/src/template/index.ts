export * from './add'
export * from './get'
export * from './query'
export * from './remove'
export * from './update'
export * from './versions'
