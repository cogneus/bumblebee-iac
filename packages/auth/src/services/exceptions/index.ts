import { StatusCodes, AppException } from 'bumblebee-common'
import { AuthExceptionCodes } from './exception-codes'

export const KeyNotFound = (key: string) =>
  new AppException({
    code: AuthExceptionCodes.KeyNotFound,
    status: StatusCodes.Forbidden,
    message: `Key with name '${key}' not found`,
    source: KeyNotFound,
  })
