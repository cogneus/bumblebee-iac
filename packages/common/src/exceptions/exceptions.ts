import { AppException } from './app-exception.class'
import { ExceptionCodes } from './exception-codes'
import { StatusCodes } from './status-codes'

export const InvalidToken = () =>
  new AppException({
    code: ExceptionCodes.InvalidToken,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidToken,
  })

export const InvalidAuthorizartion = () =>
  new AppException({
    code: ExceptionCodes.InvalidAuthorizartion,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidAuthorizartion,
  })

export const EmptyRequest = () =>
  new AppException({
    code: ExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Request data is missing',
    source: EmptyRequest,
  })

export const InvalidService = (service: string) =>
  new AppException({
    code: ExceptionCodes.InvalidService,
    status: StatusCodes.Error,
    message: `Invalid service. ${service}`,
    source: InvalidService,
  })

export const InvalidRequest = (innerMessage: string) =>
  new AppException({
    code: ExceptionCodes.InvalidRequest,
    status: StatusCodes.BadRequest,
    message: `Request data is invalid. ${innerMessage}`,
    source: InvalidRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })

export const EnvVarNotFound = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.EnvVarNotFound,
    status: StatusCodes.Error,
    message: `The following environment variables not found ${keys.join()}`,
    source: EnvVarNotFound,
  })

export const InvalidModeSource = (resourcePrefix: string, eventSourceARN: string | undefined) =>
  new AppException({
    code: ExceptionCodes.InvalidModeSource,
    status: StatusCodes.Error,
    message: `Cannot derive stage from event source. eventSourceARN: ${eventSourceARN} RESOURCE_PREFIX: ${resourcePrefix}`,
    source: InvalidModeSource,
  })
