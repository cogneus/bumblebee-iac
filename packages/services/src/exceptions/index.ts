import { StatusCodes, AppException } from 'bumblebee-common'
import { TemplateExceptionCodes } from './exception-codes'

export const TemplateNotFound = (id: string) =>
  new AppException({
    code: TemplateExceptionCodes.TemplateNotFound,
    status: StatusCodes.NotFound,
    message: `Template with id '${id}' not found`,
    source: TemplateNotFound,
  })

export const EmptyRequest = () =>
  new AppException({
    code: TemplateExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Template request data is missing',
    source: EmptyRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: TemplateExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })
