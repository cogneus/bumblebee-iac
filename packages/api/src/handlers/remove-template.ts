import { apiGatewayWrapper, warmupWrapper, standardEnvVars, Token } from 'bumblebee-common'
import { removeTemplate } from 'bumblebee-services'
import { checkAPIEnvironmentVars, checkWritePermissions, getClientOptions } from '../helpers'
import { TemplatesParams } from '../types'

checkAPIEnvironmentVars()

export const main = warmupWrapper(
  apiGatewayWrapper<any, TemplatesParams, Token, void>({
    handler: ({ params: { id }, serviceContext }) => removeTemplate(serviceContext, id!),
    ...standardEnvVars(),
    serviceName: 'template-api',
    logName: 'remove-template',
    getClientOptions,
    authCheck: checkWritePermissions,
  })
)
