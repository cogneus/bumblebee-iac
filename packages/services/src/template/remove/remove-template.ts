import { IDBClient, ServiceContext, ensureClient } from 'bumblebee-common'
import { getTemplate } from '../get/get-template'
import { MissingAttributes } from '../../exceptions'
import { createCommands } from './create-commands'
import { templateVersions } from '../versions'
import { DBItemPrimaryBase } from '../../types'

export const removeTemplate = async (context: ServiceContext, id: string): Promise<void> => {
  if (!id) {
    throw MissingAttributes(['id'])
  }
  await getTemplate(context, id)
  const db = ensureClient<IDBClient>(context, 'db')
  let nextCursor: string | undefined
  do {
    const { cursor, items } = await templateVersions<DBItemPrimaryBase>(context, id, {
      cursor: nextCursor,
      includeCurrent: true,
      fields: ['PK', 'SK'],
      limit: 20,
    })
    const params = createCommands(items, db.tables.templates.name)
    await db.transactWrite(params)
    nextCursor = cursor
  } while (nextCursor)
}
