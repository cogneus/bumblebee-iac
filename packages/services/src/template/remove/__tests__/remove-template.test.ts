import { MissingAttributes } from '../../../exceptions'
import { serviceContext } from '../../__mocks__/service-context.mock'
import { dbTemplateItem } from '../../__mocks__/template.mock'
import { removeTemplate } from '../remove-template'
import { NIL as emptyId } from 'uuid'

describe('Remove template tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  const { PK, SK } = dbTemplateItem

  test('db request should match snapshot', async () => {
    let input, queryInput
    ;(serviceContext.clients.db?.get as jest.Mock).mockImplementationOnce(() => dbTemplateItem)
    ;(serviceContext.clients.db?.transactWrite as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return null
    })
    ;(serviceContext.clients.db?.query as jest.Mock).mockImplementation((params, _, cursor) => {
      queryInput = params
      return {
        items: [
          { PK, SK },
          { PK, SK },
          { PK, SK },
        ],
        cursor: cursor ? undefined : 'test',
      }
    })
    const result = await removeTemplate(serviceContext, emptyId)
    expect({
      input,
      queryInput,
      result,
    }).toMatchSnapshot()
  })

  test('request should throw error', async () => {
    expect(async () => {
      await removeTemplate(serviceContext, null as any)
    }).rejects.toThrow(MissingAttributes(['id']))
  })
})
