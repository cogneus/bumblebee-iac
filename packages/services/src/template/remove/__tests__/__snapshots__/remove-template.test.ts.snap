// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove template tests db request should match snapshot 1`] = `
{
  "input": [
    {
      "Delete": {
        "Key": {
          "PK": "mock-PK",
          "SK": "mock-SK",
        },
        "TableName": "name-templates",
      },
    },
    {
      "Delete": {
        "Key": {
          "PK": "mock-PK",
          "SK": "mock-SK",
        },
        "TableName": "name-templates",
      },
    },
    {
      "Delete": {
        "Key": {
          "PK": "mock-PK",
          "SK": "mock-SK",
        },
        "TableName": "name-templates",
      },
    },
  ],
  "queryInput": {
    "ExpressionAttributeNames": {
      "#PK": "PK",
    },
    "ExpressionAttributeValues": {
      ":PK": "template#00000000-0000-0000-0000-000000000000",
    },
    "KeyConditionExpression": "(#PK = :PK)",
    "ProjectionExpression": "PK, SK",
    "ScanIndexForward": true,
    "TableName": "name-templates",
  },
  "result": undefined,
}
`;
